pkg/mutation/schema/schema.go

```diff
@@ -2,383 +2,179 @@ package schema
 		}
-		s.remove(mutator.Path().Nodes)
-		if s.root == nil {
+		s.Remove(id, cachedMutator.Path().Nodes)
+		db.schemas[gvk] = s
+
+		if len(s.ReferencedBy) == 0 {
+			// The root node is no longer referenced by any mutators.
 			delete(db.schemas, gvk)
 		}
 	}
-}
-
-type scheme struct {
-	gvk  schema.GroupVersionKind
-	root *node
-}
-
-func (s *scheme) add(ref []parser.Node) error {
-	if s.root == nil {
-		s.root = &node{}
-	}
-	return s.root.add(ref)
-}
-
-func (s *scheme) remove(ref []parser.Node) {
-	if s.root == nil {
-		return
-	}
-	s.root.remove(ref)
-	if s.root.referenceCount == 0 {
-		s.root = nil
-	}
-}
-
-type node struct {
-	referenceCount uint
-	nodeType       parser.NodeType
-
-	// list-type nodes have a key field and only one child
-	keyField *string
-	child    *node
-
-	// object-type nodes only have children
-	children map[string]*node
-}
-
-// backup creates a shallow copy of the node we can restore in case of error
-func (n *node) backup() *node {
-	return &node{
-		referenceCount: n.referenceCount,
-		nodeType:       n.nodeType,
-		keyField:       n.keyField,
-	}
-}
-
-func (n *node) restore(backup *node) {
-	n.referenceCount = backup.referenceCount
-	n.nodeType = backup.nodeType
-	n.keyField = backup.keyField
-}
 
-func (n *node) add(ref []parser.Node) error {
-	backup := n.backup()
-	n.referenceCount++
+	// Remove the mutator from the cache.
+	delete(db.cachedMutators, id)
 
-	// we should stop recursing when len(ref) == 1, but
-	// this guards against infinite recursion in the case
-	// where there is a bug in the switch code below
-	if len(ref) == 0 {
-		return nil
-	}
+	// This ID's conflicts are resolved since the ID no longer exists.
+	delete(db.conflicts, id)
 
-	err := func() error {
-		current := ref[0]
-		switch t := current.Type(); t {
-		case parser.ObjectNode:
-			if n.nodeType != "" && n.nodeType != parser.ObjectNode {
-				return fmt.Errorf("node type conflict: %v vs %v", n.nodeType, parser.ObjectNode)
-			}
-			n.nodeType = parser.ObjectNode
-			obj := current.(*parser.Object)
-			if n.children == nil {
-				n.children = make(map[string]*node)
-			}
-			// we stop recursing down the path right before the last element
-			// because the last element has no more type information to add
-			if len(ref) == 1 {
-				return nil
-			}
-			child, ok := n.children[obj.Reference]
-			if !ok {
-				child = &node{}
-			}
-			if err := child.add(ref[1:]); err != nil {
-				return wrapObjErr(obj, err)
-			}
-			n.children[obj.Reference] = child
-		case parser.ListNode:
-			if n.nodeType != "" && n.nodeType != parser.ListNode {
-				return fmt.Errorf("node type conflict: %v vs %v", n.nodeType, parser.ListNode)
-			}
-			n.nodeType = parser.ListNode
-			list := current.(*parser.List)
-			if n.keyField != nil && *n.keyField != list.KeyField {
-				return fmt.Errorf("key field conflict: %s vs %s", *n.keyField, list.KeyField)
+	// Check existing conflicts.
+	for conflictID := range db.conflicts {
```

Comment:
It occurs to me that we could gather a list of mutators that could potentially be cured of all conflicts as part of the remove, which would lower the complexity of removal if there are numerous conflicting mutators, where each mutator conflicts with only one other mutator.

If that seems premature, we should add a TODO comment here, in case that optimization becomes salient.

Reply:
Added TODO. Determining this set is nontrivial.